### PR TITLE
Fix multi-asic behaviour for ecnconfig

### PR DESCRIPTION
**Pull request by bktsim-arista**
#### What I did
Previously, ecnconfig -l was not behaving correctly on multi-asic devices, as the '-n' namespace option was not available, and correct namespaces were not traversed on multi-asic devices to retrieve correct results for ecnconfig. This change fixes multi-asic behaviour of CLI commands that rely on ecnconfig.

This is a part of the set of changes being pushed for https://github.com/sonic-net/sonic-buildimage/issues/15148
This change relies on helpers added in https://github.com/sonic-net/sonic-buildimage/pull/17243

#### How I did it
Added namespace option -n and used multi_asic library to implement multi_asic behaviour. Added relevant unit tests to ensure functionality.

#### How to verify it
Run unit tests, or ecnconfig commands with option `-n`.

### CODE REVIEW

**Comment by vmittal-msft** [comment]
please re-base the PR to latest

**Review by vmittal-msft — COMMENTED** [review]
please share command and it's output from multi asic platform. Also, please make sure it still works for ToR.

File: tests/mock_tables/asic1/config_db.json [hunk 1]
```diff
@@ -197,5 +197,16 @@
         "holdtime": "0",
         "asn": "65100",
         "keepalive": "0"
```
Comment (vmittal-msft) [inline tests/mock_tables/asic1/config_db.json:199]: this profile should be for lossless not lossy

File: tests/mock_tables/asic1/config_db.json [hunk 2]
```diff
@@ -242,5 +242,16 @@
     "SYSLOG_CONFIG_FEATURE|database": {
         "rate_limit_interval": "555",
         "rate_limit_burst": "55555"
+    },
+    "WRED_PROFILE|AZURE_LOSSY": {
+            "wred_green_enable":"true",
+            "wred_yellow_enable":"true",
+            "red_max_threshold":"32760",
+            "red_min_threshold":"4095",
+            "yellow_max_threshold":"32760",
```
Comment (vmittal-msft) [inline tests/mock_tables/asic1/config_db.json:251]: please remove wred on lossy

**Comment by vmittal-msft** [comment]
@bktsim-arista 
1.Is this verified on pizzabox as well ?
2. If no namespace is given, will it show for all ?

**Comment by kenneth-arista** [comment]
@arlakshm @vmittal-msft @wenyiz2021 this PR is ready for review. All tests have passed with the latest updates. It will be easier to review with whitespace difference disabled because the new static checks requires reformatting (e.g. spacing) of touched Python files.

File: tests/ecn_input/ecn_test_vectors.py [hunk 3]
```diff
@@ -18,6 +18,39 @@
...
+wred_yellow_enable       true
+yellow_min_threshold     1048576
+green_max_threshold      2097152
+green_drop_probability   5
+yellow_max_threshold     2097152
+wred_red_enable          true
+yellow_drop_probability  5
+red_drop_probability     5
+-----------------------  -------
+
```
[the remark this answers is not in the document]
Comment (arista-hpandya) [reply tests/ecn_input/ecn_test_vectors.py:52]: @vmittal-msft WRED has been removed from lossy.

File: tests/mock_tables/asic1/config_db.json (hunk 1 above)
Comment (arista-hpandya) [reply tests/mock_tables/asic1/config_db.json]: The comment is outdated, but to address it, the test vectors are added as a dummy to validate the functionality.

File: tests/mock_tables/asic1/config_db.json (hunk 2 above)
Comment (arista-hpandya) [reply tests/mock_tables/asic1/config_db.json:249]: @vmittal-msft WRED has been removed from lossy.

**Comment by arista-hpandya** [comment]
> please share command and it's output from multi asic platform. Also, please make sure it still works for ToR.

Here are the snapshots obtained after running it on a multi-asic devise:

`ecnconfig -l -n asic0`
ecnconfig -l -n asic0

`ecnconfig -l -n asic1`
ecnconfig -l -n asic1

`ecnconfig -l`
ecnconfig -l

Behaviour of queue on mutli-asic platform
multi_asic queue get and set

Behaviour of set on multi-asic platform when no namespace is specified
multi_set_all_masic

**Comment by arista-hpandya** [comment]
> @vmittal-msft 1.Is this verified on pizzabox as well ? 2. If no namespace is given, will it show for all ?

Yes, behaviour on single asic devices is preserved and when no namespace is given, it will show/set for all

File: config/main.py [hunk 4]
```diff
@@ -6388,7 +6388,8 @@ def remove_reasons(counter_name, reasons, verbose):
 @click.option('-ydrop', metavar='<yellow drop probability>', type=click.IntRange(0, 100), help="Set yellow drop probability")
 @click.option('-gdrop', metavar='<green drop probability>', type=click.IntRange(0, 100), help="Set green drop probability")
 @click.option('-v', '--verbose', is_flag=True, help="Enable verbose output")
-def ecn(profile, rmax, rmin, ymax, ymin, gmax, gmin, rdrop, ydrop, gdrop, verbose):
+@multi_asic_util.multi_asic_click_option_namespace
+def ecn(profile, rmax, rmin, ymax, ymin, gmax, gmin, rdrop, ydrop, gdrop, verbose, namespace):
```
Comment (wenyiz2021) [inline config/main.py:6392]: I'm a little confused, without adding option `-n` here, how do we use this namespace? or is it added in an above layer

Comment (arista-hpandya) [reply config/main.py:6392]: The `-n` option is added by the multi_asic_click_option_namespace. For reference here is how it does it: https://github.com/sonic-net/sonic-utilities/blob/317e649514c9b205849ffb5ea96a6a233e38290c/utilities_common/multi_asic.py#L106

Comment (kenneth-arista) [reply config/main.py:6392]: @wenyiz2021 are you asking about the decorator `@multi_asic_util.multi_asic_click_option_namespace`? Or are you asking about passing `-n` to `scripts/ecnconfig`?  

We want to add the `namepace` parameter to both the`config ecn` and `show ecn` CLI commands to teach them to operate on multi-asic platforms.

Comment (wenyiz2021) [reply config/main.py:6392]: got it now.

**Review by vmittal-msft — APPROVED** [review]
I am assuming this is verified on T0/T1 device as well to make sure it is fine.

**Comment by arista-hpandya** [comment]
> I am assuming this is verified on T0/T1 device as well to make sure it is fine.

No, it has not been tested on T0/T1 devices yet. I will update the PR once I have tested it on the same.

**Comment by arista-hpandya** [comment]
> I am assuming this is verified on T0/T1 device as well to make sure it is fine.

The changes have been verified on T0 and T1 devices, the snapshots of the same are attached below. TL;DR: It should be safe to merge

### T0
ecnconfig multi_set
ecnconfig queue test

---
### T1
ecnconfig_multi_set
ecnconfig_queue_test

**Pull request merged**